src/cluster_import.c

```diff
@@ -0,0 +1,1948 @@
+}
+
+int shouldRewriteHashtableIndex(int didx, void *privdata) {
+    return isSlotInSlotRanges(didx, (list *)privdata);
+}
+
+int childSnapshotForSyncSlot(int req, rio *rdb, void *privdata) {
+    UNUSED(req);
+    int retval = rewriteAppendOnlyFileRio(rdb, 1, shouldRewriteHashtableIndex, privdata);
+    rioWrite(rdb, "*3\r\n", 4);
+    rioWriteBulkString(rdb, "CLUSTER", 7);
+    rioWriteBulkString(rdb, "SYNCSLOTS", 9);
+    rioWriteBulkString(rdb, "SNAPSHOT-EOF", 12);
+    return retval;
+}
+
+void slotExportLinkBeginSnapshot(slotMigrationLink *link) {
+    connection **conns = zmalloc(sizeof(connection *));
+    *conns = link->client->conn;
+    serverLog(LL_NOTICE,
+              "Beginning snapshot of slot export link %.40s for slots (%s) to target %.40s.",
+              link->linkname,
+              link->slot_ranges_str,
+              link->nodename);
+    if (saveSnapshotToConnectionSockets(conns, 1, 1, 0, childSnapshotForSyncSlot, 1,
+                                        link->slot_ranges) != C_OK) {
+        serverLog(LL_WARNING,
+                  "Slot export link %.40s failed to start slot export for slots (%s)",
+                  link->linkname,
+                  link->slot_ranges_str);
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED, "Failed to start snapshot");
+        return;
+    }
+    updateSlotMigrationLinkState(link, SLOT_EXPORT_SNAPSHOTTING);
+    if (server.debug_pause_after_fork) debugPauseProcess();
+}
+
+void clusterHandleSlotExportBackgroundSaveDone(int bgsaveerr) {
+    listIter li;
+    listNode *ln;
+    listRewind(server.cluster->slot_migration_links, &li);
+    while ((ln = listNext(&li))) {
+        slotMigrationLink *link = (slotMigrationLink *)ln->value;
+        if (link->type != SLOT_MIGRATION_EXPORT) continue;
+        if (link->state != SLOT_EXPORT_SNAPSHOTTING) {
+            continue;
+        }
+        if (bgsaveerr == C_OK) {
+            serverLog(LL_NOTICE,
+                      "Finished snapshotting slots (%s) to target %.40s, beginning incremental stream...",
+                      link->slot_ranges_str,
+                      link->nodename);
+            slotExportBeginStreaming(link);
+        } else {
+            serverLog(LL_WARNING,
+                      "Failed to snapshot slots (%s) to target %.40s",
+                      link->slot_ranges_str,
+                      link->nodename);
+            finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED, "Failed to perform snapshot");
+        }
+    }
+}
+
+int clusterSlotMigrationShouldInstallWriteHandler(client *c) {
+    slotMigrationLink *link = (slotMigrationLink *)c->slot_migration_link;
+    if (!link || link->type != SLOT_MIGRATION_EXPORT) {
+        return 1;
+    }
+    return link->state != SLOT_EXPORT_SNAPSHOTTING &&
+           link->state != SLOT_EXPORT_WAITING_TO_SNAPSHOT;
+}
+
+void clusterFeedSlotExportLinks(int dbid, robj **argv, int argc) {
+    UNUSED(dbid);
```

Comment:
Given that we support multiple DBs in cluster mode now, this can't be right. I suppose we need to put `SELECT` commands into the respective stream like `replicationFeedReplicas` does when necessary (once at the beginning and when commands should work on a different DB)

Reply:
Let me make refactor this `SELECT` logic out for reuse